
```{r eval=FALSE}
library(Seurat)
library(tidyverse)
library(cowplot)
library(patchwork)
library(WGCNA)

library(Matrix)
library(viridis)
library(harmony)
library(RColorBrewer)
library(ggpubr)
library(tictoc)
library(RColorBrewer)
library(Hmisc)
library(corrplot)
library(enrichR)
library(GeneOverlap)
library(grid)
library(gridExtra)
library(igraph)
library(ggrepel)
#library(hdWGCNA)
enableWGCNAThreads(nThreads = 8)
theme_set(theme_cowplot())
set.seed(12345)


# spatial plotting functions
source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')
source("/pub/smorabit/hdWGCNA/bin/spatial_functions.R")

#devtools::install_github('smorabit/hdWGCNA', ref='dev')
library(hdWGCNA)

setwd('/dfs7/swaruplab/smorabit/analysis/scWGCNA/')

# output directories
data_dir <- "data/"
fig_dir <- 'figures/'



# re-load scWGCNA dataset:
load(file='data/Zhou_color_scheme.rda')


# reload process ed with all cell types
# seurat_obj <-  readRDS(file=paste0(data_dir, 'Zhou_scWGCNA_all_celltypes.rds'))


load('/dfs7/swaruplab/smorabit/analysis/AD_NucSeq_2019/batch_correction/liger/update/celltype-analysis/data/color_scheme.rda')
cp <- unlist(color_scheme_snRNA_celltype)
cp[names(cp) == 'EX'] <- 'turquoise'


setwd('/dfs7/swaruplab/smorabit/analysis/scWGCNA/project_modules/')

seurat_zhou <- readRDS(file= '/dfs7/swaruplab/smorabit/analysis/scWGCNA/brain_iterative/data/Zhou_scWGCNA_all_celltypes.rds')


```

Project Bulk RNA-seq modules into single-cell dataset

```{r eval=FALSE}

# load Morabito & miyoshi et al dataset
seurat_obj <- readRDS("/dfs7/swaruplab/smorabit/analysis/scWGCNA/data/Swarup_2021.rds")

# load consensus modules from
consensus_modules <- read.csv("/dfs7/swaruplab/smorabit/analysis/scWGCNA/data/hmg_cWGCNA_modules.csv")

consensus_modules <- consensus_modules %>%
  dplyr::select(c(-Ensembl.Gene.ID)) %>%
  dplyr::rename(c(gene_name = GeneSymbol, module = ModuleLabels, color = Initially.Assigned.Module.Color)) %>%
   dplyr::select(c(gene_name, module, color))

consensus_modules <- subset(consensus_modules, gene_name %in% rownames(seurat_obj))

# remove duplicate gene names
consensus_modules <- consensus_modules[match(unique(consensus_modules$gene_name), consensus_modules$gene_name),]

seurat_obj <- FindVariableFeatures(seurat_obj, nfeatures=2000)
seurat_obj <- ScaleData(seurat_obj)

seurat_obj <- ProjectModules(
  seurat_obj,
  modules = consensus_modules,
  group.by.vars = "Batch",
  seurat_ref = NULL,
  wgcna_name = "None",
  wgcna_name_proj = 'HMG_2020'
)

saveRDS(seurat_obj, file = 'data/Swarup_2021_hmg_projected.rds')


plot_list <- ModuleFeaturePlot(seurat_obj, order='shuffle', raster=TRUE, raster_dpi=400, alpha=1, restrict_range=FALSE, raster_scale=0.25)

pdf("figures/featureplot_hmg_projected.pdf",height=9, width=16)
wrap_plots(plot_list, ncol=6)
dev.off()


# get hMEs from seurat object
MEs <- GetMEs(seurat_obj, harmonized=TRUE)
mods <- colnames(MEs); mods <- mods[mods != 'grey']

# add hMEs to Seurat meta-data:
seurat_obj@meta.data <- cbind(seurat_obj@meta.data, MEs)

# plot with Seurat's DotPlot function
p <- DotPlot(seurat_obj, features=mods, group.by = 'annotation', dot.min=0.25)

# flip the x/y axes, rotate the axis labels, and change color scheme:
p <- p +
  coord_flip() +
  RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue')

# plot output
pdf("figures/dotplot_hmg_projected.pdf",height=6, width=12)
p
dev.off()


```

Set up human spatial dataset

```{r eval=FALSE}

seurat_vis <- readRDS('~/swaruplab/smorabit/analysis/scWGCNA/data/maynard_prefrontal.rds')

patch <- VisDimPlot(
  seurat_vis,
  group.by = 'layer_guess_reordered_short',
  sample_col = 'sample_name',
  dpi=600,
  ncol = 2,
  text_size=15
)

pdf(paste0(fig_dir, 'hex_ananotation_human.pdf'), width=6, height=6)
patch
dev.off()


```

Project modules into Human spatial dataset

```{r eval=FALSE}


seurat_vis <- ProjectModules(
  seurat_vis,
  seurat_ref = seurat_zhou,
  group.by.vars = "sample_name",
  gene_mapping=hg38_mm10_genes,
  genome1_col="hg38_name", # genome of reference data
  genome2_col="hg38_id", # genome of query data
  wgcna_name = "ASC",
  wgcna_name_proj = 'ASC_Projected'
)




# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, hg38_id))

hg38_mm10_genes %<>% subset(hg38_id %in% rownames(seurat_vis))

hg38_ids <- unique(hg38_mm10_genes$hg38_id)
hg38_genes <- unique(hg38_mm10_genes$hg38_name)
hg38_mm10_genes <- hg38_mm10_genes[match(hg38_genes, hg38_mm10_genes$hg38_name),]


dim(hg38_mm10_genes)
length(hg38_ids)
length(hg38_genes)




########################################
# Dot Plot
########################################


MEs <- GetMEs(seurat_vis)
modules <- GetModules(seurat_vis)
mods <- levels(modules$module)
mods <- mods[mods!='grey']

seurat_vis@meta.data <- cbind(seurat_vis@meta.data, MEs)


# make dotplot
p <- DotPlot(
  seurat_vis,
  group.by='layer_guess_reordered',
  features = rev(mods)
) + coord_flip() + RotatedAxis() +
  scale_color_gradient2(high='red', mid='grey95', low='blue') + xlab('') + ylab('') +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=1)
  )

pdf(paste0(fig_dir, 'maynard_ASC_hMEs_dotplot.pdf'), width=9, height=5)
p
dev.off()

# remove from seurat obj
seurat_vis@meta.data <- cbind(seurat_vis@meta.data, MEs)

seurat_vis@meta.data <- seurat_vis@meta.data[,!(colnames(seurat_vis@meta.data) %in% colnames(MEs))]





for(cur_mod in mods){
  print(cur_mod)
  p <- SampleFeaturePlot(
    seurat_vis,
    feature=cur_mod,
    sample_col = "sample_name",
    ncol = 4,
    raster=TRUE,
    plot_max = 'q99',
    plot_min = 0,
    colfunc = inferno,
    rev_colors=TRUE,
    dpi=400,
  )

  pdf(paste0(fig_dir, 'ME_spatial_featureplots/', cur_mod, '_featureplot.pdf'), width=10, height=3)
  print(p)
  dev.off()

}


```

Project into mouse spatial

```{r eval=FALSE}

# load 10X genomics visium dataset
seurat_vis_mouse <- readRDS("~/swaruplab/smorabit/analysis/scWGCNA/data/10x_visium_scWGCNA.rds")

image_df <- seurat_vis_mouse@images$anterior1@coordinates
seurat_vis_mouse$row <- image_df$row
seurat_vis_mouse$imagerow <- image_df$imagerow
seurat_vis_mouse$col <- image_df$col
seurat_vis_mouse$imagecol <- image_df$imagecol

# process the Visium dataset with Seurat:
seurat_vis <- seurat_vis %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA() %>%
  RunUMAP(dims=1:30) %>%
  FindNeighbors(dims=1:30) %>%
  FindClusters(res=0.75)


p1 <- DimPlot(seurat_vis_mouse, reduction = "umap", label = TRUE) +
  umap_theme() +
  NoLegend()

p2 <- SpatialDimPlot(seurat_vis_mouse, label = TRUE, label.size = 3) +
  NoLegend()

png(paste0(fig_dir, 'mouse_vis_umap.png'), units='in', res=400, width=10, height=5)
p1 | p2
dev.off()








# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))

hg38_mm10_genes <- subset(hg38_mm10_genes, mm10_name != '' & hg38_name != '')

# TODO
# need to make sure that there's only one entry for each gene in hg38_mm10_genes

mm10_genes <- unique(hg38_mm10_genes$mm10_name)
hg38_genes <- unique(hg38_mm10_genes$hg38_name)
hg38_mm10_genes <- hg38_mm10_genes[match(mm10_genes, hg38_mm10_genes$mm10_name),]



 seurat_mg <- readRDS(file='/dfs7/swaruplab/smorabit/analysis/scWGCNA/microglia/data/AD_MG_scWGCNA.rds')

seurat_vis_mouse <- ProjectModules(
  seurat_vis_mouse,
  seurat_ref = seurat_mg,
  #group.by.vars = "sample_name",
  gene_mapping=hg38_mm10_genes,
  genome1_col="hg38_name", # genome of reference data
  genome2_col="mm10_name", # genome of query data
  wgcna_name = "MG",
  wgcna_name_proj = 'MG_Projected'
)






########################################
# Dot Plot
########################################


MEs <- GetMEs(seurat_vis_mouse)
modules <- GetModules(seurat_vis_mouse)
mods <- levels(modules$module)
mods <- mods[mods!='grey']

seurat_vis_mouse@meta.data <- cbind(seurat_vis_mouse@meta.data, MEs)

#
# # make dotplot
# p <- DotPlot(
#   seurat_vis_mouse,
#   group.by='layer_guess_reordered',
#   features = rev(mods)
# ) + coord_flip() + RotatedAxis() +
#   scale_color_gradient2(high='red', mid='grey95', low='blue') + xlab('') + ylab('') +
#   theme(
#     plot.title = element_text(hjust = 0.5),
#     axis.line.x = element_blank(),
#     axis.line.y = element_blank(),
#     panel.border = element_rect(colour = "black", fill=NA, size=1)
#   )
#
# pdf(paste0(fig_dir, 'maynard_ASC_hMEs_dotplot.pdf'), width=9, height=5)
# p
# dev.off()
#
# # remove from seurat obj
# seurat_vis@meta.data <- cbind(seurat_vis@meta.data, MEs)
#
# seurat_vis@meta.data <- seurat_vis@meta.data[,!(colnames(seurat_vis@meta.data) %in% colnames(MEs))]
#




for(cur_mod in mods){
  print(cur_mod)
  p <- SampleFeaturePlot(
    seurat_vis_mouse,
    feature=cur_mod,
    sample_col = "orig.ident",
    ncol = 1,
    raster=TRUE,
    plot_max = 'q99',
    plot_min = 0,
    colfunc = inferno,
    rev_colors=TRUE,
    dpi=400,
  )

  pdf(paste0(fig_dir, 'mouse_ME_spatial_featureplots/', cur_mod, '_featureplot.pdf'), width=5, height=5)
  print(p)
  dev.off()

}



```
