---
title: "scWGCNA Basics"
output: rmarkdown::html_vignette
description: >
  Tutorial for applying the core functions of scWGCNA.
vignette: >
  %\VignetteIndexEntry{scWGCNA Basics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---




# Overview and prerequisites

This tutorial covers the basics of using scWGCNA to perform co-expression network
analysis on single-cell data. Here, we demonstrate scWGCNA using a processed
single-nucleus RNA-seq (snRNA-seq) dataset of human cortical samples from
[this publication](https://www.nature.com/articles/s41591-019-0695-9). This
dataset has already been fully processed using a standard single-cell
transcritpomics analysis pipeline such as [Seurat](https://satijalab.org/seurat/)
or [Scanpy](https://scanpy.readthedocs.io/en/stable/). If you would
like to follow this tutorial using your own dataset, you first need to satisfy the
following prerequisites:

* A single-cell or single-nucleus transcriptomics dataset in [Seurat format](https://www.rdocumentation.org/packages/Seurat/versions/3.0.1/topics/CreateSeuratObject).
* Normalize the gene expression matrix [NormalizeData](https://satijalab.org/seurat/reference/normalizedata).
* Identify highly variable genes [VariableFeatures](https://satijalab.org/seurat/reference/findvariablefeatures).
* Scale the normalized expression data [ScaleData](https://satijalab.org/seurat/reference/scaledata)
* Perform dimensionality reduction [RunPCA](https://satijalab.org/seurat/reference/runpca)
  and batch correction if needed [RunHarmony](https://www.rdocumentation.org/packages/harmony/versions/1.0/topics/RunHarmony).
* Non-linear dimensionality reduction [RunUMAP](https://satijalab.org/seurat/reference/runumap) for visualizations.
* Group cells into clusters ([FindNeighbors](https://satijalab.org/seurat/reference/findneighbors) and  [FindClusters](https://satijalab.org/seurat/reference/findclusters)).

An example of running the prerequisite data processing steps can be found in
the [Seurat Guided Clustering Tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).

## Load the dataset and required libraries

First we will load the single-cell dataset and the required R libraries for this
tutorial.


```{r eval=FALSE}

#knitr::include_graphics("man/figures/logo.png")

library(Seurat)
library(tidyverse)
library(cowplot)
library(patchwork)
library(scWGCNA)
theme_set(theme_cowplot())

# set random seed for reproducibility
set.seed(12345)

# load the Zhou et al snRNA-seq dataset
seurat_obj <- readRDS('data/Zhou_control.rds')

```

Here we will plot the non-linear dimensionality reduction (UMAP) colored by cell
type just to check that we have loaded the data correctly, and to make sure that
we have grouped cells into clusters and cell types.

```{r eval=FALSE}
p <- DimPlot(seurat_obj, group.by='cell_type', label=FALSE) +
   umap_theme + ggtitle('Cell Type')

p
```

![](figures/basic_tutorial/umap_celltype.png)


# Set up Seurat object for WGCNA

Before running scWGCNA, we first have to set up the Seurat object. Most of the
information computed by scWGCNA is stored in the Seurat object's `@misc` slot.
A single Seurat object can hold multiple scWGCNA experiments, for example representing
different cell types in the same single-cell dataset. Notably, since we consider
scWGCNA to be a downstream data analysis step, ***we do not support subsetting
the Seurat object*** after `SetupForWGCNA` has been run.

Here we will set up the Seurat object using the `SetupForWGCNA` function, specifying
the name of the scWGNCA experiment. This function also selects the genes that will be
used for WGCNA. The user can select genes using three different approaches using
the `gene_select` parameter:

* `variable`: use the genes stored in the Seurat object's `VariableFeatures`.
* `fraction`: use genes that are expressed in a certain fraction of cells for
  in the whole dataset or in each group of cells, specified by `group.by`.
* `custom`: use genes that are specified in a custom list.

In this example, we will select genes that are expressed in at least 5% of cells
in this dataset, and we will name our scWGCNA experiment "tutorial".


```{r eval=FALSE}

seurat_obj <- SetupForWGCNA(
  seurat_obj,
  gene_select = "fraction", # the gene selection approach
  fraction = 0.05, # fraction of cells that a gene needs to be expressed in order to be included
  wgcna_name = "tutorial" # the name of the scWGCNA experiment
)

```

# Construct metacells

After we have set up our Seurat object, the first step in the scWGCNA pipeine is
to construct metacells from the single-cell
dataset. Briefly, metacells are aggregates of small groups of similar cells originating
from the same biological sample of origin. Nearest neighbors algorithm is used to
identify groups of similar cells to aggregate, and then the average expression
is computed, thus yielding a metacell gene expression matrix. The sparsity of the
metacell expression matrix is considerably reduced when compared to the original
expression matrix, and therefore it is preferable to use. We were originally motivated
to use metacells in place of the original single cells because correlation network
approaches such as WGCNA are sensitive to data sparsity. Furthermore, single-cell
epigenomic approaches, such as [Cicero](https://www.cell.com/molecular-cell/pdfExtended/S1097-2765(18)30547-1), employ a similar metacell aggregation approach prior to constructing co-accessibility
networks.

scWGCNA includes a function `MetacellsByGroups` to construct metacell expression
matrices given a single-cell dataset. This function constructs a new Seurat object
which is stored internally in the scWGCNA experiment. The `group.by` parameter determines which
groups metacells will be constructed in. We only want to construct metacells from
cells that came from the same biological sample of origin, so it is critical to
pass that information to scWGCNA via the `group.by` parameter. Additionally, we usually
construct metacells for each cell type separately. Thus, in this example, we are grouping
by `Sample` and `cell_type` to achieve the desired result.

The number of cells to be aggregated `k` should be tuned based on the size of the input
dataset, in general a lower number for `k` can be used for small datasets. We generally use `k` values
between 20 and 75. The dataset used for this tutorial has 40,039 cells, ranging from
890 to 8,188 in each biological sample, and here we used `k=25`.

***Warning:*** we have found that the metacell aggregation approach does not yield good results for
extremely underrepresented cell types. For example, in this dataset, the brain vascular
cells (pericytes and endothelial cells) were the least represented, and we have
excluded them from this analysis.

Here we construct metacells and normalize the resulting expression matrix
using the following code:

```{r eval=FALSE}

# construct metacells  in each group
seurat_obj <- MetacellsByGroups(
  seurat_obj = seurat_obj,
  group.by = c("cell_type", "Sample"), # specify the columns in seurat_obj@meta.data to group by
  k = 25, # nearest-neighbors parameter
  ident.group = 'cell_type' # set the Idents of the metacell seurat object
)

# normalize metacell expression matrix:
seurat_obj <- NormalizeMetacells(seurat_obj)

```

## Optional: Process the Metacell Seurat Object

Since we store the Metacell expression information as its own Seurat object,
can we run Seurat functions on it? Absolutely! We have included a function
to easily extract the metacell seurat object.

```{r eval=FALSE}
metacell_obj <- GetMetacellObject(seurat_obj)
```

Additionally, we have included a few wrapper functions to apply the Seurat workflow
to the metacell object within the scWGCNA experiment. Here we apply these wrapper
functions to process the metacell object and visualize the aggregated expression
profiles in two dimensions with UMAP.

```{r eval=FALSE}

# process metacell seurat object with wrapper functions
seurat_obj <- NormalizeMetacells(seurat_obj) # already run above
seurat_obj <- ScaleMetacells(seurat_obj)
seurat_obj <- RunPCAMetacells(seurat_obj, features=VariableFeatures(seurat_obj))
seurat_obj <- RunHarmonyMetacells(seurat_obj, group.by.vars='Sample')
seurat_obj <- RunUMAPMetacells(seurat_obj, reduction='harmony', dims=1:15)

p1 <- DimPlotMetacells(seurat_obj, group.by='cell_type') + umap_theme + ggtitle("Cell Type")
p2 <- DimPlotMetacells(seurat_obj, group.by='Sample') + umap_theme + ggtitle("Sample")

p1 | p2

```

![](figures/basic_tutorial/umap_metacells.png)


# Co-expression network analysis

In this section we discuss how to perform co-expression network analysis with
scWGNCA on the inhibitory neuron (INH) cells in our example dataset.

## Transpose the expression matrix

Seurat objects stores gene expression matrices such that each row is a gene and
each column is a cell. However, WGCNA expects each column to be a gene, thus we
have to transpose our expression matrix. Furthermore, for this analysis, we only
want to include the inhibitory neurons, so we have to subset our expression data
prior to constructing the network. scWGCNA includes the `SetDatExpr` function to store the
transposed expression matrix for a given group of cells that will be used for downstream
network analysis. The metacell expression matrix
or the single-cell expression matrix can be used for WGCNA, by default `use_metacells=TRUE`.
This function allows the user to specify which slot to take the expression matrix from,
for example if the user wanted to apply [SCTransform](https://satijalab.org/seurat/articles/sctransform_vignette.html) normalization instead of `NormalizeData`.

The following code is used to store the transposed metacell expression matrix
for the INH cells in our Seurat object:

```{r eval=FALSE}

seurat_obj <- SetDatExpr(
  seurat_obj,
  group.by='cell_type', # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups
  group_name = "INH", # the name of the group of interest in the group.by column
  use_metacells=TRUE,
  slot = 'data'
)


```


## Select soft-power threshold

Next we will select the **"soft power threshold"**. This is an extremely important
step in the scWGNCA pipleine (and for vanilla WGCNA). Under the hood, WGCNA
constructs a correlation adjacency matrix to beging inferring co-expression
relationships. The correlations are raised to a power, which is assumed to
reduce the amount of noise present in the correlation matrix. Therefore, it is
critical to determine a proper value for the soft power threshold.

We include a function `TestSoftPowers`, to perform a parameter sweep for different
soft power thresholds. Essentially `TestSoftPowers` is a wrapper to the WGCNA
function `pickSoftThreshold`. This function helps us to guide our choice in a soft power
threshold for constructing the co-expression network by comparing. For WGCNA,
the network should ideally have a [scale-free topology](https://en.wikipedia.org/wiki/Scale-free_network),
therefore the `TestSoftPowers` function models how closely resembles a scale-free
graph at different soft power thresholds.

The following code performs the parameter sweep and outputs a summary figure.

```{r eval=FALSE}

# Test different soft powers:
seurat_obj <- TestSoftPowers(
  seurat_obj,
  setDatExpr = FALSE, # set this to FALSE since we did this above
  outfile="INH_softpowers.pdf" # output a figure to this file
)

```

![](figures/basic_tutorial/softpower.png)

The general guidance for WGCNA and scWGCNA is to pick the lowest soft power threshold
that has a Scale Free Topology Model Fit greater than or equal to 0.8, so in this case
we would select our soft power threshold as 9.

Notably, the output table from the parameter sweep is stored in the scWGCNA experiment
and can be accessed using the `GetPowerTable` function:

```{r eval=FALSE}
power_table <- GetPowerTable(seurat_obj)
```

## Construct co-expression network

It's the moment you've been waiting for, we now have everything that we need to
construct our co-expression network. Here we use the scWGCNA function `ConstructNetwork`,
which calls the WGCNA function [`blockwiseConsensusModules`](https://www.rdocumentation.org/packages/WGCNA/versions/1.70-3/topics/blockwiseConsensusModules) under the hood. This function has
quite a few parameters to play with if you are an advanced user, but we have selected
default parameters that work well with many single-cell datasets. The parameters for
`blockwiseConsensusModules` can be passed directly to `ConstructNetwork` with the
same parameter names.

The following code construtcts the co-expression network using the soft power threshold
selected above:

```{r eval=FALSE}

# construct co-expression network:
seurat_obj <- ConstructNetwork(
  seurat_obj, soft_power=9,
  setDatExpr=FALSE
)

```

scWGCNA also includes a function `PlotDendrogram` to visualize the WGCNA dendrogram, a common visualization
to show the different co-expression modules resulting from the network analysis.
Each leaf on the dendrogram represents a single gene, and the color at the bottom
indicates the co-expression module assignment.

```{r eval=FALSE}

PlotDendrogram(seurat_obj, main='INH scWGCNA Dendrogram')

```

![](figures/basic_tutorial/inh_dendro.png)

# Module Eigengenes and Intramodular Connectivity

In this section we will cover how to compute module eigengenes in single cells,
and how to compute the intramodular connectivity for each gene.

## Compute harmonized module eigengenes

Module Eigengenes (MEs) are a commonly used metric to summarize the gene expression profile
of an entire co-expression module. Briefly, module eigengenes are computed by performing
principal component analysis (PCA) on the subset of the gene expression matrix
comprising each module. The first PC of each of these PCA matrices are the MEs.

Dimensionality reduction techniques are a very hot topic in single-cell genomics.
It is well known that technical artifacts can muddy the analysis of single-cell
datasets, and over the years there have been many methods that aim to reduce the
effects of these artifacts. Therefore it stands to reason that MEs would be subject to
these technical artifacts as well, and scWGCNA seeks to alleviate these effects.

scWGCNA includes a function `ModuleEigengenes` to compute MEs in the Seurat object
using Seurat's `RunPCA` function under the hood. Additionally, we allow the user
to apply Harmony batch correction to alleviate the technical effects.

As in the Seurat workflow, the expression matrix must be scaled with `ScaleData`
before we run `ModuleEigengenes`. The following code scales the expression matrix
for the WGCNA genes, and then performs the module eigengene computation harmonizing
by the Sample of origin using the `group.by.vars` parameter.

```{r eval=FALSE}

# expression matrix for all the WGCNA genes:
seurat_obj <- Seurat::ScaleData(
 seurat_obj,
 features = GetWGCNAGenes(seurat_obj),
 vars.to.regress = c('n_genes_by_counts', 'pct_counts_mt')
)

# compute all MEs in the full single-cell dataset
seurat_obj <- ModuleEigengenes(
 seurat_obj,
 group.by.vars="Sample"
)

```

## Compute hub gene signature scores

```{r eval=FALSE}

seurat_obj <- ModuleExprScore(
  seurat_obj,
  n_genes = 25,
  method='Seurat'
)

```

## Visualization

```{r eval=FALSE}

plot_list <- ModuleFeaturePlot(seurat_obj, order=TRUE)
wrap_plots(plot_list, ncol=6)




```

## Module Correlation analysis

```{r eval=FALSE}

ModuleCorrelogram(seurat_obj, sig.level = 0.001, pch.cex=2)

```
