---
title: "Vignettes overview"
output: rmarkdown::html_vignette
description: >
  Document showing an overview of the different analysis vignettes.
vignette: >
  %\VignetteIndexEntry{Vignettes overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Co-expression network analysis

These tutorials cover the essentials of performing co-expression network
analysis in single-cell transcriptomics data, and visualizing the key results.

### [hdWGCNA basics](basic_tutorial.html)
This tutorial covers the essential functions to construct a co-expression network
in single-cell transcriptomics data with hdWGCNA.


<img src="figures/basic_tutorial/Zhou_featureplot_hMEs_selected_wide.png" width="500" height="500">


### [Network visualization](network_visualizations.html)

This tutorial highlights several approaches for visualizing the hdWGCNA co-expression
networks.

<img src="figures/network_vis/hubgene_umap_igraph.png" width="400" height="400">

## Biological context for co-expression modules

These tutorials will provide further biological context for our co-expression
modules, potentially revealing what experimental conditions and biological
processes that these modules are involved in.

### [Differential module eigengene (DME) analysis](differential_MEs.html)

This tutorial covers how to compare module eigengenes between experimental groups.

<img src="figures/DMEs/test_DME_volcano.png" width="400" height="400" href="differential_MEs.html">


### [Module trait correlation](module_trait_correlation.html)

This tutorial covers how to correlate continuous and categorical variables with
module eigengenes or module expression scores, revealing which modules are related
to different experimental conditions or covariates.

<img src="figures/mt_correlation/ME_Trait_correlation_fdr.png" width="400" height="400" href="module_trait_correlation.html">

### [Enrichment analysis](enrichment_analysis.html)

This tutorial shows how to use Enrichr to compare the gene members of each co-expression
module to curated gene lists, thereby pointing towards the biological functions
of the co-expression modules.

<img src="figures/enrichment/GO_dotplot.png" width="400" height="400">

## Exploring modules in external datasets

### [Projecting modules to new datasets](projecting_modules.html)

This tutorial covers how to project co-expression modules from a reference
to a query dataset.

<img src="figures/projection/compare_umaps.png" width="500" height="500">

### [Module preservation and reproducibility](module_preservation.html)

This tutorial covers statistical methods for assessing the preservation and
reproducibility of co-expression networks using external datasets.


<img src="figures/projection/module_preservation_summary.png" width="500" height="500">

### [Cross-species and cross-modality analysis](projecting_modules_cross.html)

This tutorial covers how to project co-expression modules from a reference
to a query dataset for special cases where the data modality or the species do
not match between the reference and the query.

<img src="figures/projection/atac_umap_covplot.png" width="700" height="700">

## Advanced topics

### [Consensus network analysis](consensus_wgcna.html)


### [Motif analysis](motif_analysis.html)

## Other

### [Module customization](customization.html)

This tutorial covers how to change the default names and colors for hdWGCNA modules.

<img src="figures/customization/featureplot.png" width="400" height="400">


### [Using SCTransform normalized data](sctransform.html)

This tutorial covers how to use SCTransform normalized data in hdWGCNA.

<img src="figures/sctransform/dendro_SCT_compare.png" width="400" height="400" href="sctransform.html">
