---
title: "Module preservation and reproducibility"
output: rmarkdown::html_vignette
description: >
  Tutorial for performing module preservation tests in hdWGCNA.
vignette: >
  %\VignetteIndexEntry{Module preservation and reproducibility}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Data-driven models are thought to be useful when they are generalizable across
different datasets. If we think about co-expression networks from this perspective,
we need a way to statistically quantify the conservation of co-expression modules
that were identified in one datasets across external datasets. This is a common theme
in machine learning, where a model is learned on a training dataset but evaluated
using an external validation dataset.

In the [previous tutorial](projecting_modules.html), we projected the co-expression
modules from a reference to a query dataset, but we did not quantify the degree
to which these modules were preserved across datasets. In this tutorial, we perform statistical tests to estimate the conservation of
projected modules.

First we must load the data and the required libraries:
```{r eval=FALSE}

# single-cell analysis package
library(Seurat)

# plotting and data science packages
library(tidyverse)
library(cowplot)
library(patchwork)

# co-expression network analysis packages:
library(WGCNA)
library(hdWGCNA)

# network analysis & visualization package:
library(igraph)

# using the cowplot theme for ggplot
theme_set(theme_cowplot())

# set random seed for reproducibility
set.seed(12345)

# load the Zhou et al snRNA-seq dataset
seurat_ref <- readRDS('data/Zhou_control.rds')

# load the Morabito & Miyoshi 2021 snRNA-seq dataset
seurat_query <- readRDS(file=paste0(data_dir, 'Morabito_Miyoshi_2021_control.rds'))

```

## Module preservation analysis

In their 2011 paper titled
["Is My Network Module Preserved and Reproducible?"](https://doi.org/10.1371/journal.pcbi.1001057), Langfelder et al discuss statistical methods for
module preservation analysis in co-expression network analysis.  Notably, module preservation analysis can be used
to assess the reproducibility of co-expression networks, but it can also be used
for specific biological analyses, for example to identify which modules are
significantly preserved across different disease conditions, tissues, developmental
stages, or even evolutionary time.

The first step in module preservation analysis is to project modules from a reference to a query dataset, which is explained in more detail in the [previous tutorial](projecting_modules.html).

```{r eval=FALSE}

seurat_query <- ProjectModules(
  seurat_obj = seurat_query,
  seurat_ref = seurat_ref,
  wgcna_name = "INH",
  wgcna_name_proj="INH_projected",
  assay="RNA" # assay for query dataset
)

```

Next, we set up the expression matrices for the query and reference datasets. Either the 
single-cell or the metacell gene expression matrices can be used here by setting the `use_metacells` 
flag in `SetDatExpr`. In general we recommend using the same matrix that was used to construct the network 
for the reference (in this case, the metacell matrix). 


```{r eval=FALSE}

# set expression matrix for reference dataset
seurat_ref <- SetDatExpr(
  seurat_ref,
  group_name = "INH",
  group.by = "cell_type"
)

# set expression matrix for query dataset:
seurat_query <- SetDatExpr(
  seurat_query,
  group_name = "INH",
  group.by = "cell_type",
  use_metacells = FALSE
)

```

<details> <summary> ** Using the metacell expression matrix for the query dataset ** </summary>

If you choose to use the metacell expression matrix for the query dataset, you have to first run `MetacellsByGroups` on the query dataset.

```{r eval=FALSE}

# compute metacells
seurat_query <- MetacellsByGroups(
  seurat_obj = seurat_query,
  group.by = c("cell_type", "Sample"), 
  reduction = 'harmony',
  k = 25,
  max_shared = 10, 
  ident.group = 'cell_type'
)
seurat_query <- NormalizeMetacells(seurat_query)

# set expression matrix for query dataset:
seurat_query <- SetDatExpr(
  seurat_query,
  group_name = "INH",
  group.by = "cell_type"
)

```

</details>

Now we can run the `ModulePreservation` function. Note that this function does take a while to run since it is permutation based, but it can be sped up by lowering the `n_permutations` parameter, but we do not recommend any lower than 100 permutations. Also note that the `parallel` option currently does not work.

```{r eval=FALSE}

# run module preservation function
seurat_query <- ModulePreservation(
  seurat_query,
  seurat_ref = seurat_ref,
  name="Zhou-INH",
  verbose=3,
  n_permutations=250
)

# getthe module preservation table
mod_pres <- GetModulePreservation(seurat_query, "Zhou-INH")$Z
obs_df <- GetModulePreservation(seurat_query, "Zhou-INH")$obs

grep('summary', colnames(mod_pres))
grep('summary', colnames(obs))

```

Please consult [this tutorial](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/ModulePreservation/Tutorials/)
for an explanation of the different stats. Currently working on writing up a
description of the stats on this page. The visualizations are inspired by
those included in the [original WGCNA module preservation tutorials](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/ModulePreservation/Tutorials/MiniTutorial-MouseLiver.pdf).


## Visualize preservation stats

In this section, we use the function `PlotModulePreservation` to visualize the
statistics we computed in the previous section. This function generates a scatter
plot showing the module size versus the module preservation stats.

```{r eval=FALSE}

plot_list <- PlotModulePreservation(
  seurat_query,
  name="Zhou-INH",
  statistics = "summary"
)

wrap_plots(plot_list, ncol=2)


```

<img src="figures/projection/module_preservation_summary.png" width="700" height="700">

Plot the ranking stats

```{r eval=FALSE}

# plot ranking stats
plot_list <- PlotModulePreservation(
  seurat_query,
  name="Zhou-INH",
  statistics = "rank"
)


png(paste0(fig_dir, 'module_preservation_rank.png'), width=10, height=10, res=400, units='in')
wrap_plots(plot_list, ncol=2)
dev.off()

```

<img src="figures/projection/module_preservation_rank.png" width="700" height="700">


Plot all of the different stats:

```{r eval=FALSE}

plot_list <- PlotModulePreservation(
  seurat_query,
  name="Zhou-INH",
  statistics = "all",
  plot_labels=FALSE
)

wrap_plots(plot_list, ncol=6)

```

<img src="figures/projection/module_preservation_all.png" width="700" height="700">
