---
title: "Motif Analysis"
output: rmarkdown::html_vignette
description: >
  Tutorial for identifying transcription factor binding motifs that are related
  to scWGCNA modules.
vignette: >
  %\VignetteIndexEntry{Motif Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Load the snRNA-seq data and the required libraries:
```{r eval=FALSE}

# single-cell analysis package
library(Seurat)

# plotting and data science packages
library(tidyverse)
library(cowplot)
library(patchwork)

# co-expression network analysis packages:
library(WGCNA)
library(scWGCNA)

# network analysis & visualization package:
library(igraph)

# packages for TF motif analysis
library(JASPAR2020)
library(motifmatchr)
library(TFBSTools)
library(EnsDb.Hsapiens.v86)
library(GenomicRanges)

# using the cowplot theme for ggplot
theme_set(theme_cowplot())

# set random seed for reproducibility
set.seed(12345)

# load the Zhou et al snRNA-seq dataset
seurat_ref <- readRDS('data/Zhou_control.rds')

```

Run Motif scan:

```{r eval=FALSE}

# get the pfm from JASPAR2020 using TFBSTools
pfm_core <- TFBSTools::getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# run the motif scan with these settings for the mouse dataset
seurat_obj <- MotifScan(
  seurat_obj,
  species_genome = 'hg38',
  pfm = pfm_core,
  EnsDb = EnsDb.Hsapiens.v86
)
dim(GetMotifMatrix(seurat_obj))



# TF target genes
target_genes <- GetMotifTargets(seurat_obj)

```

Overlap between TF target genes and co-expression modules

```{r eval=FALSE}


# overlap between modules & TF target genes:
seurat_obj<- OverlapModulesMotifs(seurat_obj)
head(GetMotifOverlap(seurat_obj))

# plot the top TFs overlapping with
MotifOverlapBarPlot(
  seurat_obj,
  #motif_font = 'xkcd_regular',
  outdir = 'motifs/MotifOverlaps/',
  plot_size=c(5,6)
)


```

Compute motif target gene expression score:

```{r eval=FALSE}

library(UCell)
seurat_obj <- MotifTargetScore(
  seurat_obj,
  method='UCell'
)


```

Plot overlap between motif of interest and modules:

```{r eval=FALSE}


df <- GetMotifOverlap(seurat_obj)

cur_df <- df %>% subset(tf == 'NR4A2')

plot_var <- 'odds_ratio'
p <- cur_df %>%
  ggplot(aes(y=reorder(module, odds_ratio), x=odds_ratio)) +
  geom_bar(stat='identity', fill=cur_df$color) +
  geom_vline(xintercept = 1, linetype='dashed', color='gray') +
  geom_text(aes(label=Significance), color='black', size=3.5, hjust='center') +
  ylab('') +
  xlab("Odds Ratio") +
  theme(
    plot.title = element_text(hjust = 0.5)
  ) +


pdf(paste0(fig_dir, 'Nr4a2_motif_overlap_or.pdf'), width=3, height=3, useDingbats=FALSE)
p
dev.off()

```
