---
title: "Projecting modules"
output: rmarkdown::html_vignette
description: >
  Tutorial for applying the core functions of scWGCNA.
vignette: >
  %\VignetteIndexEntry{Projecting modules}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

In this tutorial, we begin to analyze the co-expresion modules detected in one
dataset in external datasets. Rather than building a new co-expression network
from scratch in a new dataset, we can take the modules from one dataset and project
them into the new dataset. Before starting this tutorial, make sure that you
have constructed the co-expression network as in the [scWGCNA basics](articles/basics_tutorial.html). The main scWGCNA tutorials have been using the control
brain samples from [this publication](https://www.nature.com/articles/s41591-019-0695-9),
and now we will project the inhibitory neruon modules from Zhou et al into the
control brain samples from [Morabito and Miyoshi 2021](https://doi.org/10.1038/s41588-021-00894-z).

First we must load the data and the required libraries:
```{r eval=FALSE}

# single-cell analysis package
library(Seurat)

# plotting and data science packages
library(tidyverse)
library(cowplot)
library(patchwork)

# co-expression network analysis packages:
library(WGCNA)
library(scWGCNA)

# network analysis & visualization package:
library(igraph)

# using the cowplot theme for ggplot
theme_set(theme_cowplot())

# set random seed for reproducibility
set.seed(12345)

# load the Zhou et al snRNA-seq dataset
seurat_ref <- readRDS('data/Zhou_control.rds')

# load the Morabito & Miyoshi 2021 snRNA-seq dataset
seurat_query <- readRDS(file=paste0(data_dir, 'Morabito_Miyoshi_2021_control.rds'))

```

## Projecting modules

In this section we project the modules from the Zhou et al inhibitory neuron
scWGCNA experiment into the Morabito & Miyoshi et al control brain dataset. We
refer to the Zhou et al dataset as the "reference" dataset, and the Morabito & Miyoshi
et al dataset as the "query" dataset. Just [as we had done before](basic_tutorial.html) when building the co-expression network from scratch, the basic single-cell pipeline
has to be 0done on the query dataset (normalization, scaling, variable features, PCA, batch correction, UMAP, clustering).
First we make a UMAP plot to visualize the two datasets to ensure they have both
been processed.

<details> <summary> Code </summary>
```{r eval=FALSE}

p1 <- DimPlot(seurat_ref, group.by='cell_type', label=TRUE) +
   umap_theme() +
   ggtitle('Zhou') +
   NoLegend()

p2 <- DimPlot(seurat_query, group.by='cell_type', label=TRUE) +
   umap_theme() +
   ggtitle('Morabito & Miyoshi') +
   NoLegend()

p1 | p2

```
</details>

<img src="figures/projection/compare_umaps.png" width="600" height="600">

Next we will run the function `ProjectModules` to project the modules from
the reference dataset into the query dataset. If the genes used for
co-expression network analysis in the reference dataset have not been scaled
in the query dataset, Seurat's `ScaleData` function will be run from within
`ProjectModules`. We perform the following analysis steps to project the modules into the query dataset:

* Run `ProjectModules` to compute the module eigengenes in the query dataset
  based on the gene modules in the reference dataset.
* Optionally run `ModuleExprScore` to compute hub gene expression scores for
  the projected modules.
* Run `ModuleConnectivity` to compute intramodular connectivity (*kME*) in the
  query dataset.

```{r eval=FALSE}

# Project modules from query to reference dataset
seurat_query <- ProjectModules(
  seurat_obj = seurat_query,
  seurat_ref = seurat_ref,
  scale_genes = TRUE,
  # vars.to.regress = c(), # optionally regress covariates when running ScaleData
  wgcna_name_proj="Zhou_projected", # name of the new scWGCNA experiment in the query dataset
  wgcna_name = "tutorial" # name of the scWGCNA experiment in the ref dataset
)

```


As you can see it only takes running one function to project co-expression modules
from one dataset into another using scWGCNA. Optionally, we can also compute the
hub gene expression scores and the intramodular connectivity for the projected modules.
Note that if we do not run the `ModuleConnectivity` function on the query dataset,
the *kME* values in the module assignment table `GetModules(seurat_query)` are the
*kME* values from the reference dataset.


<details> <summary> Code </summary>
```{r eval=FALSE}
# compute module hub gene expression scores for projected modules:
seurat_query <- ModuleExprScore(
  seurat_query,
  n_genes = 25,
  method='Seurat'
)

# compute intramodular connectivity
seurat_query <- ModuleConnectivity(seurat_query, assay="RNA", slot="data")

```
</details>

We can extract the projected module eigengenes using the `GetMEs` function.

```{r eval=FALSE}

projected_hMEs <- GetModules(seurat_query)

```

The projected modules can be used in most of the downstream scWGCNA analysis
tasks and visualization functions, such as [module trait correlation](module_trait_correlation.html),
or.

## Visualization

Compare the hub gene UMAP between the projected and the query (use gganimate!!)
