<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Integrating co-expression and protein-protein interaction networks • hdWGCNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integrating co-expression and protein-protein interaction networks">
<meta name="description" content="Integrating co-expression and protein-protein interaction networks
">
<meta property="og:description" content="Integrating co-expression and protein-protein interaction networks
">
<meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hdWGCNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.04</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/basic_tutorial.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><h6 class="dropdown-header" data-toc-skip>Core functionality</h6></li>
    <li><a class="dropdown-item" href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a></li>
    <li><a class="dropdown-item" href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a></li>
    <li><a class="dropdown-item" href="../articles/isoform_pbmcs.html">Isoform co-expression network analysis with PacBio MAS-Seq</a></li>
    <li><a class="dropdown-item" href="../articles/network_visualizations.html">Network visualization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Biological context for co-expression modules</h6></li>
    <li><a class="dropdown-item" href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a></li>
    <li><a class="dropdown-item" href="../articles/module_trait_correlation.html">Module trait correlation</a></li>
    <li><a class="dropdown-item" href="../articles/enrichment_analysis.html">Enrichment analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Exploring modules in external datasets</h6></li>
    <li><a class="dropdown-item" href="../articles/projecting_modules.html">Projecting modules to new datasets</a></li>
    <li><a class="dropdown-item" href="../articles/module_preservation.html">Module preservation and reproducibility</a></li>
    <li><a class="dropdown-item" href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Advanced topics</h6></li>
    <li><a class="dropdown-item" href="../articles/tf_network.html">Transcription factor regulatory network analysis</a></li>
    <li><a class="dropdown-item" href="../articles/ppi_integration.html">Integrating co-expression and protein-protein interaction networks</a></li>
    <li><a class="dropdown-item" href="../articles/consensus_wgcna.html">Consensus network analysis</a></li>
    <li><a class="dropdown-item" href="../articles/pseudobulk.html">hdWGCNA with pseudobulk data</a></li>
    <li><a class="dropdown-item" href="../articles/pseudotime.html">Co-expression module dynamics with pseudotime</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other</h6></li>
    <li><a class="dropdown-item" href="../articles/customization.html">Module customization</a></li>
    <li><a class="dropdown-item" href="../articles/other_metacells.html">Alternative metacell algorithms</a></li>
    <li><a class="dropdown-item" href="../articles/sctransform.html">Using SCTransform normalized data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">All vignettes</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/smorabit/hdWGCNA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Integrating co-expression and protein-protein interaction networks</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/smorabit/hdWGCNA/blob/HEAD/vignettes/ppi_integration.Rmd" class="external-link"><code>vignettes/ppi_integration.Rmd</code></a></small>
      <div class="d-none name"><code>ppi_integration.Rmd</code></div>
    </div>

    
    
<p>Compiled: 24-01-2024</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Co-expression networks are a useful approach for identifying biological patterns in a transcriptomic dataset in an unsupervised manner. However, interpreting individual co-expression patterns is not always straightforward. What does a strong co-expression between a pair of genes mean? Does gene A regulate gene B? Are they part of the same pathway? Expressed in the same cell type, disease state, timepoint, etc? In this tutorial, we aim to add more interpretability to the co-expression network by integrating it with a <strong>protein-protein interaction (PPI) network</strong>. This <em>advanced tutorial</em> is inspired by some of our previous works in bulk transcriptomics for <a href="https://academic.oup.com/hmg/article/29/17/2899/5892988" class="external-link">human</a> and <a href="https://www.nature.com/articles/s41591-018-0223-3" class="external-link">mouse</a> datasets, but we have updated the code for compatibility with hdWGCNA.</p>
<p><strong>Important Note:</strong> Before running the code in this tutorial, you must first perform the standard co-expression network analysis and module identification with hdWGCNA, as shown in our <a href="basic_tutorial.html">other tutorial</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># single-cell analysis package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plotting and data science packages:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># co-expression network analysis packages:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" class="external-link">WGCNA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://smorabit.github.io/hdWGCNA/">hdWGCNA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># network analysis &amp; visualization package:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using the cowplot theme for ggplot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set random seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># re-load the Zhou et al snRNA-seq dataset</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'data/Zhou_control.rds'</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="install-additional-packages">Install additional packages<a class="anchor" aria-label="anchor" href="#install-additional-packages"></a>
</h3>
<p>For this analysis, we rely on the Bioconductor package <code>STRINGdb</code> to retrieve the PPI data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install STRINGdb</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'STRINGdb'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load STRINGdb</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">STRINGdb</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="integrating-co-expression-and-ppi-networks">Integrating co-expression and PPI networks<a class="anchor" aria-label="anchor" href="#integrating-co-expression-and-ppi-networks"></a>
</h2>
<p>Since we already have a co-expression network generated with hdWGCNA, now we need a PPI network. First, we use the R package <code>STRINGdb</code> to set up a PPI network. Here we briefly explain some of the options for initializing the network, but please consult the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/STRINGdb/inst/doc/STRINGdb.pdf" class="external-link">STRINBdb documentation</a> for further details about using this package. In principle, other PPI resources may be used instead of <code>STRINGdb</code>.</p>
<div class="section level3">
<h3 id="create-the-ppi-igraph-object-">Create the PPI <code>igraph</code> object.<a class="anchor" aria-label="anchor" href="#create-the-ppi-igraph-object-"></a>
</h3>
<p>In this section, we represent the PPI network using an <code>igraph</code> object. First, we generate a <code>STRINGdb</code> object using the class constructor <code>STRINGdb$new</code>.</p>
<ul>
<li>
<code>version</code>: the STRING version.</li>
<li>
<code>species</code>: the species taxonomy ID. Human is 9606, mouse is 10090. For other species, <a href="http://www.ncbi.nlm.nih.gov/taxonomy" class="external-link">check the NCBI taxonomoy</a>.</li>
<li>
<code>score_threshold</code>: the minimum combined interaction score for a PPI to be included in the network.</li>
<li>
<code>network_type</code>: the type of network interactions to use. STRING uses both <code>functional</code> and <code>physical</code> interaction data to rank PPIs for the <code>full</code> network, but you can choose to just use one or the other if you wish.</li>
</ul>
<p>After we initialize the <code>STRINGdb</code> object, we call the function <code>get_graph()</code> to convert it to an <code>igraph</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># initialize the stringdb</span></span>
<span><span class="va">string_db</span> <span class="op">&lt;-</span> <span class="va">STRINGdb</span><span class="op">$</span><span class="fu">new</span><span class="op">(</span> </span>
<span>    version<span class="op">=</span><span class="st">"12.0"</span>, species<span class="op">=</span><span class="fl">9606</span>, </span>
<span>    score_threshold<span class="op">=</span><span class="fl">200</span>, </span>
<span>    network_type<span class="op">=</span><span class="st">"full"</span>, </span>
<span>    input_directory<span class="op">=</span><span class="st">""</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># retrieve the full PPI graph as an igraph object</span></span>
<span><span class="va">ppi</span> <span class="op">&lt;-</span> <span class="va">string_db</span><span class="op">$</span><span class="fu">get_graph</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Now we have <code>ppi</code>, an <code>igraph</code> object containing the full network of PPIs. Let’s quickly check how many protiens (vertices) and how many PPIs (edges) are in this network.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_ppis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n_proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"N proteins: "</span>, <span class="va">n_proteins</span>, <span class="st">' | N PPIs: '</span>, <span class="va">n_ppis</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "N proteins: 19611 | N PPIs: 7533072"</code></pre>
</div>
<div class="section level3">
<h3 id="match-protein-and-gene-names">Match protein and gene names<a class="anchor" aria-label="anchor" href="#match-protein-and-gene-names"></a>
</h3>
<p>Let’s check how the proteins are named in the PPI network.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "9606.ENSP00000000233" "9606.ENSP00000000412" "9606.ENSP00000001008"
[4] "9606.ENSP00000001146" "9606.ENSP00000002125" "9606.ENSP00000002165"</code></pre>
<p>Currently, the proteins are named using their Ensembl protein IDs with the species taxonomy ID (9606) appended to the front. In order to integrate the PPI network and the co-expression network, we need a common naming convention. To solve this, we will convert these protein IDs to gene names using <a href="https://www.ensembl.org/info/data/biomart/index.html" class="external-link">Ensembl BioMart</a>. Using BioMart, we generated a table with the following attributes:</p>
<ul>
<li>Gene.stable.ID</li>
<li>Gene.stable.ID.version</li>
<li>Transcript.stable.ID</li>
<li>Transcript.stable.ID.version</li>
<li><strong>Protein.stable.ID</strong></li>
<li>Protein.stable.Id.version</li>
<li><strong>Gene.name</strong></li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the table from BioMart</span></span>
<span><span class="va">ens_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">'data/ensembl_Grch38_gene_protein.txt'</span>, sep<span class="op">=</span><span class="st">'\t'</span>, header<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ens_df</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="13%">
<col width="15%">
<col width="14%">
<col width="18%">
<col width="13%">
<col width="17%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th>Gene.stable.ID</th>
<th>Gene.stable.ID.version</th>
<th>Transcript.stable.ID</th>
<th>Transcript.stable.ID.version</th>
<th>Protein.stable.ID</th>
<th>Protein.stable.ID.version</th>
<th>Gene.name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ENSG00000210049</td>
<td>ENSG00000210049.1</td>
<td>ENST00000387314</td>
<td>ENST00000387314.1</td>
<td></td>
<td></td>
<td>MT-TF</td>
</tr>
<tr class="even">
<td>ENSG00000211459</td>
<td>ENSG00000211459.2</td>
<td>ENST00000389680</td>
<td>ENST00000389680.2</td>
<td></td>
<td></td>
<td>MT-RNR1</td>
</tr>
<tr class="odd">
<td>ENSG00000210077</td>
<td>ENSG00000210077.1</td>
<td>ENST00000387342</td>
<td>ENST00000387342.1</td>
<td></td>
<td></td>
<td>MT-TV</td>
</tr>
<tr class="even">
<td>ENSG00000210082</td>
<td>ENSG00000210082.2</td>
<td>ENST00000387347</td>
<td>ENST00000387347.2</td>
<td></td>
<td></td>
<td>MT-RNR2</td>
</tr>
<tr class="odd">
<td>ENSG00000209082</td>
<td>ENSG00000209082.1</td>
<td>ENST00000386347</td>
<td>ENST00000386347.1</td>
<td></td>
<td></td>
<td>MT-TL1</td>
</tr>
<tr class="even">
<td>ENSG00000198888</td>
<td>ENSG00000198888.2</td>
<td>ENST00000361390</td>
<td>ENST00000361390.2</td>
<td>ENSP00000354687</td>
<td>ENSP00000354687.2</td>
<td>MT-ND1</td>
</tr>
</tbody>
</table>
<p>Our Seurat object and the co-expression network use gene names, so we will convert the protein names in the PPI network to gene names.</p>
<p><strong>IMPORTANT NOTE:</strong> This approach is sufficient for the purpose of the tutorial, but ideally we should convert the gene names in the Seurat object to Ensembl Gene IDs, and then match the Ensembl Protein IDs with the Ensembl Gene IDs because Ensembl IDs are more stable/robust than gene names.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remove the Taxonomy ID from the protein names</span></span>
<span><span class="va">ppi</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/set_vertex_attr.html" class="external-link">set_vertex_attr</a></span><span class="op">(</span></span>
<span>    <span class="va">ppi</span>, <span class="st">'name'</span>, </span>
<span>    value<span class="op">=</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span>, <span class="st">'9606.'</span>, <span class="st">''</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset the table for genes that are in the Seurat obj &amp; have a corresponding </span></span>
<span><span class="co"># protein entry in the PPI network</span></span>
<span><span class="va">ens_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>    <span class="va">ens_df</span>, </span>
<span>    <span class="va">Gene.name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>    <span class="va">Protein.stable.ID</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset the PPI network to just proteins found in our Ensembl data</span></span>
<span><span class="va">ppi</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/subgraph.html" class="external-link">subgraph</a></span><span class="op">(</span></span>
<span>    <span class="va">ppi</span>,</span>
<span>    vids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ens_df</span><span class="op">$</span><span class="va">Protein.stable.ID</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># match the protein names w/ gene names, update the PPI network</span></span>
<span><span class="va">ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span>, <span class="va">ens_df</span><span class="op">$</span><span class="va">Protein.stable.ID</span><span class="op">)</span></span>
<span><span class="va">ppi</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/set_vertex_attr.html" class="external-link">set_vertex_attr</a></span><span class="op">(</span></span>
<span>    <span class="va">ppi</span>, <span class="st">'name'</span>, </span>
<span>    value<span class="op">=</span><span class="va">ens_df</span><span class="op">$</span><span class="va">Gene.name</span><span class="op">[</span><span class="va">ix</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "ARF5"    "M6PR"    "FKBP4"   "CYP26B1" "NDUFAF7" "FUCA2"  </code></pre>
</div>
<div class="section level3">
<h3 id="combine-the-co-expression-and-ppi-networks">Combine the co-expression and PPI networks<a class="anchor" aria-label="anchor" href="#combine-the-co-expression-and-ppi-networks"></a>
</h3>
<p>Now that we have converted the protein IDs to gene names, we subset the PPI to match the same genes as the co-expression network. We also subset the co-expression network (Topological Overlap Matrix, TOM) to match the PPI.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># keep genes that are in the co-expression network and in the PPI network</span></span>
<span><span class="va">keep_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetWGCNAGenes.html">GetWGCNAGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">keep_genes</span> <span class="op">&lt;-</span> <span class="va">keep_genes</span><span class="op">[</span><span class="va">keep_genes</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">ppi</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># subset the PPI network</span></span>
<span><span class="va">ppi</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/subgraph.html" class="external-link">subgraph</a></span><span class="op">(</span></span>
<span>    <span class="va">ppi</span>,</span>
<span>    vids <span class="op">=</span> <span class="va">keep_genes</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the adjacency matrix representation of the PPI and</span></span>
<span><span class="co"># binarize the matrix so 0 means no PPI, 1 means PPI </span></span>
<span><span class="va">ppi_mat</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/as_adjacency_matrix.html" class="external-link">as_adjacency_matrix</a></span><span class="op">(</span><span class="va">ppi</span>, sparse<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ppi_mat</span><span class="op">[</span><span class="va">ppi_mat</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># get the coex network (TOM), </span></span>
<span><span class="va">TOM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTOM.html">GetTOM</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">TOM</span> <span class="op">&lt;-</span> <span class="va">TOM</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ppi_mat</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ppi_mat</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Finally, we will perform an element-wise multiplication of the co-expression network by the binarized PPI network. This means the co-expression information will be retained only if there is evidence of a PPI for a particular gene pair.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform element-wise multiplicaiton of the TOM by the PPIs network </span></span>
<span><span class="co"># to only retain gene-gene links supported by PPIs</span></span>
<span><span class="va">TOM_ppi</span> <span class="op">&lt;-</span> <span class="va">TOM</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">ppi_mat</span><span class="op">)</span></span></code></pre></div>
<p>We can visually compare the original co-expression network with the integrated network using the function <code>ModuleTopologyHeatmap</code>. This function simply plots a heatmap of the network adjacency matrix for a particular module.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get modules and from the seurat obj</span></span>
<span><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetModules.html">GetModules</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">module</span> <span class="op">!=</span> <span class="st">'grey'</span> <span class="op">&amp;</span> <span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">keep_genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>module <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">module</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">modules</span><span class="op">$</span><span class="va">module</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get module colors for plotting </span></span>
<span><span class="va">mod_colors</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">modules</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">module</span>, <span class="va">color</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">mod_cp</span> <span class="op">&lt;-</span> <span class="va">mod_colors</span><span class="op">$</span><span class="va">color</span>; <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">mod_colors</span><span class="op">$</span><span class="va">module</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select the module to plot</span></span>
<span><span class="va">cur_mod</span> <span class="op">&lt;-</span> <span class="st">'INH-M9'</span></span>
<span></span>
<span><span class="co"># plot the original co-expression network </span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleTopologyHeatmap.html">ModuleTopologyHeatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    mod <span class="op">=</span> <span class="va">cur_mod</span>,</span>
<span>    matrix <span class="op">=</span> <span class="va">TOM</span>,</span>
<span>    matrix_name <span class="op">=</span> <span class="st">'TOM'</span>,</span>
<span>    order_by <span class="op">=</span> <span class="st">'kME'</span>,</span>
<span>    high_color <span class="op">=</span> <span class="va">mod_cp</span><span class="op">[</span><span class="va">cur_mod</span><span class="op">]</span>,</span>
<span>    type <span class="op">=</span> <span class="st">'unsigned'</span>,</span>
<span>    return_genes <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># select this option to get the gene order</span></span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># get the plot from the output list</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">tmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Co-expression only'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the gene order from the output list</span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="va">tmp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># plot the integrated network</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ModuleTopologyHeatmap.html">ModuleTopologyHeatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat_obj</span>,</span>
<span>    mod <span class="op">=</span> <span class="va">cur_mod</span>,</span>
<span>    matrix <span class="op">=</span> <span class="va">TOM_ppi</span>,</span>
<span>    matrix_name <span class="op">=</span> <span class="st">'TOM'</span>,</span>
<span>    order_by <span class="op">=</span> <span class="st">'kME'</span>,</span>
<span>    high_color <span class="op">=</span> <span class="va">mod_cp</span><span class="op">[</span><span class="va">cur_mod</span><span class="op">]</span>,</span>
<span>    type <span class="op">=</span> <span class="st">'unsigned'</span>,</span>
<span>    plot_max <span class="op">=</span> <span class="va">tmp</span><span class="op">$</span><span class="va">plot_max</span>,</span>
<span>    genes_order <span class="op">=</span> <span class="va">gene_list</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Co-expression + PPI'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine with patchwork</span></span>
<span><span class="op">(</span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span><span class="va">cur_mod</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<center>
<img src="figures/ppi_network/topology_compare.png" width="1200" height="600">
</center>
<p>In this heatmap, we can easily see that many of the gene-gene edges in the co-expression network are missing in the integrated network. This is expected, and it is because the integrated network only contains gene-gene edges where there is evidence of a protein-protein interaction.</p>
</div>
</div>
<div class="section level2">
<h2 id="identifying-hub-genes-in-the-integrated-network">Identifying hub genes in the integrated network<a class="anchor" aria-label="anchor" href="#identifying-hub-genes-in-the-integrated-network"></a>
</h2>
<p>Network hub genes are genes with strong connections in a co-expression module, and they are typically interesting candidates for further analysis. However, since we now have a different network only containing gene-gene links with PPI support, the connectivity of the network has changed substntially, and therefore the hub genes have likely changed as well.</p>
<p>In this section, we calculate the connectivity of each gene in the integrated network for one module, and compare this to the eigengene-based connectivity (kME) which we typically use to identify module hub genes. We will then make a scatter plot comparing these two measures of connectivity for each gene, and we will label the top hub genes with evidence from both the original co-expression network and the integrated network.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the list of genes in this module </span></span>
<span><span class="va">cur_genes</span> <span class="op">&lt;-</span> <span class="va">modules</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">module</span> <span class="op">==</span> <span class="va">cur_mod</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># get the kME for this module</span></span>
<span><span class="va">kme_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetHubGenes.html">GetHubGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span>, n_hubs<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">cur_genes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kme_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">kme_df</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># subset TOM for genes in this module</span></span>
<span><span class="va">cur_TOM_ppi</span> <span class="op">&lt;-</span> <span class="va">TOM_ppi</span><span class="op">[</span><span class="va">cur_genes</span>,<span class="va">cur_genes</span><span class="op">]</span> </span>
<span><span class="va">cur_TOM_ppi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">lower.tri</a></span><span class="op">(</span><span class="va">cur_TOM_ppi</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span> </span>
<span></span>
<span><span class="co"># cast from from matrix to long dataframe, remove self-links</span></span>
<span><span class="va">cur_TOM_df</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">cur_TOM_ppi</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Var1</span> <span class="op">!=</span> <span class="va">Var2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate degree for each gene and arrange by degree</span></span>
<span><span class="va">ppi_degree</span> <span class="op">&lt;-</span> <span class="va">cur_TOM_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Var1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>degree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>weighted_degree <span class="op">=</span> <span class="va">degree</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">degree</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">degree</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>gene_name <span class="op">=</span> <span class="va">Var1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ppi_degree</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ppi_degree</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># combine into 1 table:</span></span>
<span><span class="va">kme_df</span><span class="op">$</span><span class="va">ppi_degree</span> <span class="op">&lt;-</span> <span class="va">ppi_degree</span><span class="op">[</span><span class="va">kme_df</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'degree'</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#------------------------------------------------------#</span></span>
<span><span class="co"># Plot kME vs PPI connectivity</span></span>
<span><span class="co">#------------------------------------------------------#</span></span>
<span></span>
<span><span class="co"># what are the top n genes (hub genes) by kME vs by ppi? </span></span>
<span><span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">25</span></span>
<span><span class="va">top_kme</span> <span class="op">&lt;-</span> <span class="va">kme_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by<span class="op">=</span><span class="va">kME</span>, n<span class="op">=</span><span class="va">n_genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="va">top_ppi</span> <span class="op">&lt;-</span> <span class="va">kme_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by<span class="op">=</span><span class="va">ppi_degree</span>, n<span class="op">=</span><span class="va">n_genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># get the genes common to both groups</span></span>
<span><span class="va">top_both</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">top_kme</span>, <span class="va">top_ppi</span><span class="op">)</span></span>
<span><span class="va">df_kme</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">kme_df</span>, <span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_kme</span> <span class="op">&amp;</span> <span class="op">!</span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_both</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_ppi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">kme_df</span>, <span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_ppi</span> <span class="op">&amp;</span> <span class="op">!</span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_both</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_both</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">kme_df</span>, <span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_both</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define the x and y intercepts for lines showing the hub gene cutoff</span></span>
<span><span class="va">v_line</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">df_kme</span><span class="op">$</span><span class="va">kME</span><span class="op">)</span></span>
<span><span class="va">h_line</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">df_ppi</span><span class="op">$</span><span class="va">ppi_degree</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># other plotting parameters</span></span>
<span><span class="va">point_size</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">cur_color</span> <span class="op">&lt;-</span> <span class="va">mod_cp</span><span class="op">[</span><span class="va">cur_mod</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># scatter plot comparing connectivity</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">kme_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">top_kme</span>, <span class="va">top_ppi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kME</span>, y <span class="op">=</span> <span class="va">ppi_degree</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">cur_color</span>, size<span class="op">=</span><span class="va">point_size</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept<span class="op">=</span><span class="va">v_line</span>, color<span class="op">=</span><span class="st">'lightgrey'</span>, linetype <span class="op">=</span> <span class="st">'dashed'</span>, size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept<span class="op">=</span><span class="va">h_line</span>, color<span class="op">=</span><span class="st">'lightgrey'</span>, linetype <span class="op">=</span> <span class="st">'dashed'</span>, size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">df_kme</span>, size<span class="op">=</span><span class="va">point_size</span>,</span>
<span>        fill <span class="op">=</span> <span class="va">cur_color</span>, color <span class="op">=</span> <span class="va">cur_color</span>, shape <span class="op">=</span> <span class="fl">25</span>,</span>
<span>        size<span class="op">=</span><span class="va">point_size</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">df_ppi</span>, size<span class="op">=</span><span class="va">point_size</span>,</span>
<span>        fill <span class="op">=</span> <span class="va">cur_color</span>, color <span class="op">=</span> <span class="va">cur_color</span>, shape <span class="op">=</span> <span class="fl">24</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">df_both</span>, size<span class="op">=</span><span class="va">point_size</span><span class="op">*</span><span class="fl">1.5</span>,</span>
<span>        fill <span class="op">=</span> <span class="va">cur_color</span>, color<span class="op">=</span><span class="st">'black'</span>, shape<span class="op">=</span><span class="fl">23</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">df_both</span>,</span>
<span>        mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">gene_name</span><span class="op">)</span>, color<span class="op">=</span><span class="st">'black'</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Intramodular connectivity (kME)'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Intramodular PPI connectivity (degree)'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        panel.border<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">'black'</span>, fill<span class="op">=</span><span class="cn">NA</span>, linewidth<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>,</span>
<span>        axis.line.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.line.y<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span> </span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<center>
<img src="figures/ppi_network/degree_compare.png" width="1200" height="600">
</center>
<p>This plot shows the genes that are considered hub genes for the original co-expression network (bottom right, downwards triangle), in the integrated network (upper left, upwards triangle), and in both networks (upper right, diamonds). Among 25 hub genes for each network type, there are 6 in common. This approach effectively narrows down the genes of interest for a particular module by requiring external support for connectivity based on an orthogonal data source, in this case the <code>STRINGdb</code> PPI network.</p>
</div>
<div class="section level2">
<h2 id="visualizing-the-integrated-network">Visualizing the integrated network<a class="anchor" aria-label="anchor" href="#visualizing-the-integrated-network"></a>
</h2>
<p>Here we use <code>tidygraph</code> and <code>ggraph</code> to visualize the integrated network. This section is similar to our <a href="network_visualizations.html">network visualization tutorial</a>, please consult this tutorial for more information about custom network plots.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cast the network from wide to long format</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="va">cur_TOM_ppi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>gene1 <span class="op">=</span> <span class="va">Var1</span>, gene2 <span class="op">=</span> <span class="va">Var2</span>, weight<span class="op">=</span><span class="va">value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html" class="external-link">as_tbl_graph</a></span><span class="op">(</span>directed<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add the module name to the graph:</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">module</span> <span class="op">&lt;-</span> <span class="va">modules</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,<span class="st">'module'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add the hub gene status and connectivity</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">hub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_both</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">""</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">kME</span> <span class="op">&lt;-</span> <span class="va">kme_df</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'kME'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">ppi_degree</span> <span class="op">&lt;-</span> <span class="va">kme_df</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'ppi_degree'</span><span class="op">]</span></span>
<span></span>
<span><span class="va">lay</span> <span class="op">&lt;-</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">create_layout</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://r.igraph.org/reference/layout_nicely.html" class="external-link">layout_nicely</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lay</span> <span class="op">&lt;-</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">create_layout</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://r.igraph.org/reference/layout_with_graphopt.html" class="external-link">layout_with_graphopt</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></span><span class="op">(</span><span class="va">lay</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html" class="external-link">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fu"><a href="../reference/scale01.html">scale01</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span><span class="op">)</span>, color<span class="op">=</span><span class="st">'darkgrey'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">module</span>, size<span class="op">=</span><span class="va">ppi_degree</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span></span>
<span>    data<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">lay</span>, <span class="va">hub</span> <span class="op">!=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">module</span>, size<span class="op">=</span><span class="va">ppi_degree</span><span class="op">)</span>, color<span class="op">=</span><span class="st">'black'</span>, shape<span class="op">=</span><span class="fl">21</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html" class="external-link">scale_edge_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_text.html" class="external-link">geom_node_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">hub</span><span class="op">)</span>, repel<span class="op">=</span><span class="cn">TRUE</span>, max.overlaps<span class="op">=</span><span class="cn">Inf</span>, fontface<span class="op">=</span><span class="st">'italic'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cur_mod</span>, <span class="st">" integrated network"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<center>
<img src="figures/ppi_network/INH-M9_ppi_graph.png" width="1200" height="600">
</center>
<p>We can repeat this process for all modules in the co-expression network.</p>
<details><summary>
Click to see code
</summary><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">cur_mod</span> <span class="kw">in</span> <span class="va">mods</span><span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cur_mod</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># get the list of genes in this module </span></span>
<span>    <span class="va">cur_genes</span> <span class="op">&lt;-</span> <span class="va">modules</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">module</span> <span class="op">==</span> <span class="va">cur_mod</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span>    <span class="co"># get the kME for this module</span></span>
<span>    <span class="va">kme_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetHubGenes.html">GetHubGenes</a></span><span class="op">(</span><span class="va">seurat_obj</span>, n_hubs<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">cur_genes</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kme_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">kme_df</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span>    <span class="co"># subset TOM for genes in this module</span></span>
<span>    <span class="va">cur_TOM_ppi</span> <span class="op">&lt;-</span> <span class="va">TOM_ppi</span><span class="op">[</span><span class="va">cur_genes</span>,<span class="va">cur_genes</span><span class="op">]</span> </span>
<span>    <span class="va">cur_TOM_ppi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">lower.tri</a></span><span class="op">(</span><span class="va">cur_TOM_ppi</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span> </span>
<span></span>
<span>    <span class="co"># cast from from matrix to long dataframe, remove self-links</span></span>
<span>    <span class="va">cur_TOM_df</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">cur_TOM_ppi</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Var1</span> <span class="op">!=</span> <span class="va">Var2</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># calculate degree for each gene and arrange by degree</span></span>
<span>    <span class="va">ppi_degree</span> <span class="op">&lt;-</span> <span class="va">cur_TOM_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Var1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>degree <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>weighted_degree <span class="op">=</span> <span class="va">degree</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">degree</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">degree</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>gene_name <span class="op">=</span> <span class="va">Var1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ppi_degree</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ppi_degree</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span>    <span class="co"># combine into 1 table:</span></span>
<span>    <span class="va">kme_df</span><span class="op">$</span><span class="va">ppi_degree</span> <span class="op">&lt;-</span> <span class="va">ppi_degree</span><span class="op">[</span><span class="va">kme_df</span><span class="op">$</span><span class="va">gene_name</span>, <span class="st">'degree'</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co"># what are the top n genes (hub genes) by kME vs by ppi? </span></span>
<span>    <span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">25</span></span>
<span>    <span class="va">top_kme</span> <span class="op">&lt;-</span> <span class="va">kme_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by<span class="op">=</span><span class="va">kME</span>, n<span class="op">=</span><span class="va">n_genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span>    <span class="va">top_ppi</span> <span class="op">&lt;-</span> <span class="va">kme_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by<span class="op">=</span><span class="va">ppi_degree</span>, n<span class="op">=</span><span class="va">n_genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">gene_name</span></span>
<span>    <span class="va">top_both</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">top_kme</span>, <span class="va">top_ppi</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># cast the network from wide to long format</span></span>
<span>    <span class="va">graph</span> <span class="op">&lt;-</span> <span class="va">cur_TOM_ppi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>gene1 <span class="op">=</span> <span class="va">Var1</span>, gene2 <span class="op">=</span> <span class="va">Var2</span>, weight<span class="op">=</span><span class="va">value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>        <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html" class="external-link">as_tbl_graph</a></span><span class="op">(</span>directed<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># add the module name to the graph:</span></span>
<span>    <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">module</span> <span class="op">&lt;-</span> <span class="va">modules</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,<span class="st">'module'</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co"># add the hub gene status and connectivity</span></span>
<span>    <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">hub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_both</span>, <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">""</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">kME</span> <span class="op">&lt;-</span> <span class="va">kme_df</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'kME'</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">ppi_degree</span> <span class="op">&lt;-</span> <span class="va">kme_df</span><span class="op">[</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, <span class="st">'ppi_degree'</span><span class="op">]</span></span>
<span></span>
<span>    <span class="va">lay</span> <span class="op">&lt;-</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">create_layout</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://r.igraph.org/reference/layout_nicely.html" class="external-link">layout_nicely</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">lay</span> <span class="op">&lt;-</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">create_layout</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://r.igraph.org/reference/layout_with_graphopt.html" class="external-link">layout_with_graphopt</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></span><span class="op">(</span><span class="va">lay</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html" class="external-link">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fu"><a href="../reference/scale01.html">scale01</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span><span class="op">)</span>, color<span class="op">=</span><span class="st">'darkgrey'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">module</span>, size<span class="op">=</span><span class="va">ppi_degree</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span></span>
<span>        data<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">lay</span>, <span class="va">hub</span> <span class="op">!=</span> <span class="st">''</span><span class="op">)</span>, </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="va">module</span>, size<span class="op">=</span><span class="va">ppi_degree</span><span class="op">)</span>, color<span class="op">=</span><span class="st">'black'</span>, shape<span class="op">=</span><span class="fl">21</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html" class="external-link">scale_edge_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="va">mod_cp</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_text.html" class="external-link">geom_node_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">hub</span><span class="op">)</span>, repel<span class="op">=</span><span class="cn">TRUE</span>, max.overlaps<span class="op">=</span><span class="cn">Inf</span>, fontface<span class="op">=</span><span class="st">'italic'</span>, size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cur_mod</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="va">plot_list</span><span class="op">[[</span><span class="va">cur_mod</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">p</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, ncol<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</details><center>
<img src="figures/ppi_network/ppi_graph.png" width="1200" height="900">
</center>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Here we have discussed how to provide further support for the gene-gene links in the co-expression network by integrating with a protein-protein interaction network. If you are interested in further exporing specific gene-gene interactions, there are several external packages that can be explored. For example, both <a href="https://github.com/tpq/propr" class="external-link"><code>propr</code></a> and <a href="https://github.com/yelabucsf/scrna-parameter-estimation" class="external-link"><code>memento</code></a> are packages that can perform differential analysis of specific gene pairs.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
