<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ConstructMetacells — ConstructMetacells • hdWGCNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="ConstructMetacells — ConstructMetacells"><meta property="og:description" content="This function takes a Seurat object and constructs averaged 'metacells' based
on neighboring cells."><meta property="og:image" content="https://smorabit.github.io/hdWGCNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hdWGCNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.27</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/hdWGCNA.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Core functionality</li>
    <li>
      <a href="../articles/basic_tutorial.html">hdWGCNA in single-cell data</a>
    </li>
    <li>
      <a href="../articles/ST_basics.html">hdWGCNA in spatial transcriptomics data</a>
    </li>
    <li>
      <a href="../articles/isoform_pbmcs.html">Isoform co-expression network analysis with PacBio MAS-Seq</a>
    </li>
    <li>
      <a href="../articles/network_visualizations.html">Network visualization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Biological context for co-expression modules</li>
    <li>
      <a href="../articles/differential_MEs.html">Differential module eigengene (DME) analysis</a>
    </li>
    <li>
      <a href="../articles/module_trait_correlation.html">Module trait correlation</a>
    </li>
    <li>
      <a href="../articles/enrichment_analysis.html">Enrichment analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Exploring modules in external datasets</li>
    <li>
      <a href="../articles/projecting_modules.html">Projecting modules to new datasets</a>
    </li>
    <li>
      <a href="../articles/module_preservation.html">Module preservation and reproducibility</a>
    </li>
    <li>
      <a href="../articles/projecting_modules_cross.html">Cross-species and cross-modality analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced topics</li>
    <li>
      <a href="../articles/consensus_wgcna.html">Consensus network analysis</a>
    </li>
    <li>
      <a href="../articles/pseudobulk.html">hdWGCNA with pseudobulk data</a>
    </li>
    <li>
      <a href="../articles/pseudotime.html">Co-expression module dynamics with pseudotime</a>
    </li>
    <li>
      <a href="../articles/motif_analysis.html">Motif analysis</a>
    </li>
    <li>
      <a href="../articles/other_metacells.html">Alternative metacell algorithms</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/customization.html">Module customization</a>
    </li>
    <li>
      <a href="../articles/sctransform.html">Using SCTransform normalized data</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/index.html">All vignettes</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ConstructMetacells</h1>
    
    <div class="hidden name"><code>ConstructMetacells.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes a Seurat object and constructs averaged 'metacells' based
on neighboring cells.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ConstructMetacells</span><span class="op">(</span>
  <span class="va">seurat_obj</span>,
  name <span class="op">=</span> <span class="st">"agg"</span>,
  ident.group <span class="op">=</span> <span class="st">"seurat_clusters"</span>,
  k <span class="op">=</span> <span class="fl">25</span>,
  reduction <span class="op">=</span> <span class="st">"pca"</span>,
  dims <span class="op">=</span> <span class="cn">NULL</span>,
  assay <span class="op">=</span> <span class="st">"RNA"</span>,
  cells.use <span class="op">=</span> <span class="cn">NULL</span>,
  slot <span class="op">=</span> <span class="st">"counts"</span>,
  meta <span class="op">=</span> <span class="cn">NULL</span>,
  return_metacell <span class="op">=</span> <span class="cn">FALSE</span>,
  mode <span class="op">=</span> <span class="st">"average"</span>,
  max_shared <span class="op">=</span> <span class="fl">15</span>,
  target_metacells <span class="op">=</span> <span class="fl">1000</span>,
  max_iter <span class="op">=</span> <span class="fl">5000</span>,
  verbose <span class="op">=</span> <span class="cn">FALSE</span>,
  wgcna_name <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seurat_obj</dt>
<dd><p>A Seurat object</p></dd>
<dt>name</dt>
<dd><p>A string appended to resulting metalcells. Default = 'agg'</p></dd>
<dt>k</dt>
<dd><p>Number of nearest neighbors to aggregate. Default = 50</p></dd>
<dt>reduction</dt>
<dd><p>A dimensionality reduction stored in the Seurat object. Default = 'umap'</p></dd>
<dt>dims</dt>
<dd><p>A vector represnting the dimensions of the reduction to use. Either specify the names of the dimensions or the indices. Default = NULL to include all dims.</p></dd>
<dt>assay</dt>
<dd><p>Assay to extract data for aggregation. Default = 'RNA'</p></dd>
<dt>slot</dt>
<dd><p>Slot to extract data for aggregation. Default = 'counts'</p></dd>
<dt>return_metacell</dt>
<dd><p>Logical to determine if we return the metacell seurat object (TRUE), or add it to the misc in the original Seurat object (FALSE). Default to FALSE.</p></dd>
<dt>mode</dt>
<dd><p>determines how to make gene expression profiles for metacells from their constituent single cells. Options are "average" or "sum".</p></dd>
<dt>max_shared</dt>
<dd><p>the maximum number of cells to be shared across two metacells</p></dd>
<dt>target_metacells</dt>
<dd><p>the maximum target number of metacells to construct</p></dd>
<dt>max_iter</dt>
<dd><p>the maximum number of iterations in the metacells bootstrapping loop</p></dd>
<dt>verbose</dt>
<dd><p>logical indicating whether to print additional information</p></dd>
<dt>wgcna_name</dt>
<dd><p>name of the WGCNA experiment</p></dd>
</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">ConstructMetacells</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> function (seurat_obj, name = "agg", ident.group = "seurat_clusters", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     k = 25, reduction = "pca", dims = NULL, assay = "RNA", cells.use = NULL, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     slot = "counts", meta = NULL, return_metacell = FALSE, mode = "average", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     max_shared = 15, target_metacells = 1000, max_iter = 5000, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     verbose = FALSE, wgcna_name = NULL) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (is.null(wgcna_name)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         wgcna_name &lt;- seurat_obj@misc$active_wgcna</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!(reduction %in% names(seurat_obj@reductions))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stop(paste0("Invalid reduction (", reduction, "). Reductions in Seurat object: ", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             paste(names(seurat_obj@reductions), collapse = ", ")))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!(assay %in% names(seurat_obj@assays))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stop(paste0("Invalid assay (", assay, "). Assays in Seurat object: ", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             paste(names(seurat_obj@assays), collapse = ", ")))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!(slot %in% c("counts", "data", "scale.data"))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         stop(paste0("Invalid slot (", slot, "). Valid options for slot: counts, data, scale.data "))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!is.null(cells.use)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seurat_full &lt;- seurat_obj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seurat_obj &lt;- seurat_obj[, cells.use]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     reduced_coordinates &lt;- as.data.frame(seurat_obj@reductions[[reduction]]@cell.embeddings)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     reduc_orig &lt;- reduced_coordinates</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!is.null(dims)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (is.numeric(dims)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             if (!all(dims %in% 1:ncol(reduced_coordinates))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 stop("Invalid selection for dims. Check that the selected dims match the columns of the selected reduction.")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             if (!all(dims %in% colnames(reduced_coordinates))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 stop("Invalid selection for dims. Check that the selected dims match the column names of the selected reduction.")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         reduced_coordinates &lt;- reduced_coordinates[, dims]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     nn_map &lt;- FNN::knn.index(reduced_coordinates, k = (k - 1))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     row.names(nn_map) &lt;- row.names(reduced_coordinates)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     nn_map &lt;- cbind(nn_map, seq_len(nrow(nn_map)))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     good_choices &lt;- seq_len(nrow(nn_map))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     choice &lt;- sample(seq_len(length(good_choices)), size = 1, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         replace = FALSE)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     chosen &lt;- good_choices[choice]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     good_choices &lt;- good_choices[good_choices != good_choices[choice]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     it &lt;- 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     k2 &lt;- k * 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     get_shared &lt;- function(other, this_choice) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         k2 - length(union(cell_sample[other, ], this_choice))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     while (length(good_choices) &gt; 0 &amp; length(chosen) &lt; target_metacells &amp; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         it &lt; max_iter) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         it &lt;- it + 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         choice &lt;- sample(seq_len(length(good_choices)), size = 1, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             replace = FALSE)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         new_chosen &lt;- c(chosen, good_choices[choice])</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         good_choices &lt;- good_choices[good_choices != good_choices[choice]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cell_sample &lt;- nn_map[new_chosen, ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         others &lt;- seq_len(nrow(cell_sample) - 1)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         this_choice &lt;- cell_sample[nrow(cell_sample), ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         shared &lt;- sapply(others, get_shared, this_choice = this_choice)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (max(shared) &lt;= max_shared) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             chosen &lt;- new_chosen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     shared_old &lt;- shared</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     cell_sample &lt;- nn_map[chosen, ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     cells_merged &lt;- apply(cell_sample, 1, function(x) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         paste0(colnames(seurat_obj)[x], collapse = ",")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     })</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     combs &lt;- tryCatch({</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         combn(nrow(cell_sample), 2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }, error = function(cond) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         return(NA)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     })</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (any(is.na(combs))) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         warning("Metacell failed")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         return(NULL)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     shared &lt;- apply(combs, 2, function(x) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         k2 - length(unique(as.vector(cell_sample[x, ])))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     })</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (verbose) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         message(paste0("Overlap QC metrics:\nCells per bin: ", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             k, "\nMaximum shared cells bin-bin: ", max(shared), </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             "\nMean shared cells bin-bin: ", mean(shared), "\nMedian shared cells bin-bin: ", </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             median(shared)))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (mean(shared)/k &gt; 0.1) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             warning("On average, more than 10% of cells are shared between paired bins.")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     exprs_old &lt;- GetAssayData(seurat_obj, assay = assay, slot = slot)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mask &lt;- sapply(seq_len(nrow(cell_sample)), function(x) seq_len(ncol(exprs_old)) %in% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         cell_sample[x, , drop = FALSE])</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mask &lt;- Matrix::Matrix(mask)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     new_exprs &lt;- (exprs_old %*% mask)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (mode == "average") {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         new_exprs &lt;- new_exprs/k</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     colnames(new_exprs) &lt;- paste0(name, "_", 1:ncol(new_exprs))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     rownames(cell_sample) &lt;- paste0(name, "_", 1:ncol(new_exprs))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     colnames(cell_sample) &lt;- paste0("knn_", 1:ncol(cell_sample))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     metacell_obj &lt;- CreateSeuratObject(counts = new_exprs, assay = assay)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (slot == "scale.data") {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         metacell_obj &lt;- SeuratObject::SetAssayData(metacell_obj, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             slot = slot, assay = assay, new.data = as.matrix(new_exprs))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     metacell_obj$cells_merged &lt;- as.character(cells_merged)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     max_shared &lt;- max(shared)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     median_shared &lt;- median(shared)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     mean_shared &lt;- mean(shared)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     new_exprs[new_exprs &gt; 0] &lt;- 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     density &lt;- sum(Matrix::colSums(new_exprs)/(nrow(new_exprs) * </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         ncol(new_exprs)))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     run_stats &lt;- data.frame(name = name, max_shared = max_shared, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         mean_shared = mean_shared, median_shared = median_shared, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         density = density, n = ncol(new_exprs))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     metacell_obj@misc$run_stats &lt;- run_stats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (!is.null(meta)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         meta_names &lt;- names(meta)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         for (x in meta_names) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             metacell_obj@meta.data[[x]] &lt;- meta[[x]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else (warning("meta not found"))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     if (return_metacell) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         out &lt;- metacell_obj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     else {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         if (!is.null(cells.use)) {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             seurat_obj &lt;- seurat_full</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seurat_obj &lt;- SetMetacellObject(seurat_obj, metacell_obj, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             wgcna_name)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         seurat_obj &lt;- SetWGCNAParams(seurat_obj, params = list(metacell_k = k, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             metacell_reduction = reduction, metacell_slot = slot, </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             metacell_assay = assay), wgcna_name)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         out &lt;- seurat_obj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     out</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;bytecode: 0x7ff6589f4038&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;environment: namespace:hdWGCNA&gt;</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smorabit.github.io" class="external-link">Sam Morabito</a>, <a href="https://swaruplab.bio.uci.edu/" class="external-link">Swarup Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

